### DISCUSSION TITLE
Airflow 2.7.0: Webserver errors with "Worker was sent SIGTERM!"

### DISCUSSION THREAD

**Discussion by VladimirYushkevich** [apache/airflow#34340] [General] [answered]
### Apache Airflow version

Other Airflow 2 version (please specify below)

### What happened

Following the upgrade of Airflow from version 2.6.2 to 2.7.0, the `webserver` began reporting errors related to the receipt of a SIGTERM signal from the worker. This issue was not observed in earlier versions.

### What you think should happen instead

Here an example of log entry from `webserver`:
```
"name": "gunicorn.error", "filename": "glogging.py", "module": "glogging", "message": "Worker (pid:40) was sent SIGTERM!", "processName": "MainProcess", "timestamp": "2023-09-13 10:37:19,466", "severity": "ERROR"
```
Used K8s resources for CPU and Memory for all airflow pods are always ~50%.
I also don't see any correlation with scheduled tasks(we don't have many tasks). We also don't have any CPU/Memory intensive tasks.

### How to reproduce

Use the following in values.yaml when deploying the helm chart on GKE cluster:
```
executor: CeleryExecutor

airflowVersion: "2.7.0"
```

### Operating System

Debian GNU/Linux 11 (bullseye)

### Deployment

Official Apache Airflow Helm Chart

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template! If you are willing to raise PR to address this issue please do so, no need to wait for approval.

**Comment by potiuk** [comment]
It's not reproducible on our K8S installation. In order to help you to diagnose the problem you have, we need to know who is sending the signal. Somewhere likely something in your K8S installation stops webserver by seding it SIGTERM. And in order to know it, we need to know what and why is sending it. It's quite impossible to diagnose it without it.

**Comment by VladimirYushkevich** [comment]
Thank you for your response. I initially believed that "Worker" in the logs referred to an Airflow `worker` pod, but it appears to be related to some Linux internals, likely indicating something like a process holder. I see that there is no harmful impact of this signal(like stopping or restarting of `webserver`).
I still didn't find what is actually sending it. But I've noticed that just before the error occurs, there is a health check operation taking place:
```
"2023-09-13 16:43:15,991 [INFO] 172.16.0.3 - - [13/Sep/2023:16:43:15 +0000] \"GET /health HTTP/1.1\" 200 353 \"-\" \"GoogleHC/1.0\""
"2023-09-13 16:43:16,179 [INFO] 172.16.0.28 - - [13/Sep/2023:16:43:16 +0000] \"GET /health HTTP/1.1\" 200 353 \"-\" \"GoogleHC/1.0\""
"2023-09-13 16:43:16,179 [INFO] 172.16.0.28 - - [13/Sep/2023:16:43:16 +0000] \"GET /health HTTP/1.1\" 200 353 \"-\" \"GoogleHC/1.0\""
"2023-09-13 16:43:16,438 [INFO] 172.16.0.82 - - [13/Sep/2023:16:43:16 +0000] \"GET /health HTTP/1.1\" 200 353 \"-\" \"GoogleHC/1.0\""
"2023-09-13 16:43:16,438 [INFO] 172.16.0.82 - - [13/Sep/2023:16:43:16 +0000] \"GET /health HTTP/1.1\" 200 353 \"-\" \"GoogleHC/1.0\""
"2023-09-13 16:43:17,651 [INFO] 172.16.0.82 - - [13/Sep/2023:16:43:17 +0000] \"GET /health HTTP/1.1\" 200 353 \"-\" \"GoogleHC/1.0\""
"2023-09-13 16:43:17,651 [INFO] 172.16.0.82 - - [13/Sep/2023:16:43:17 +0000] \"GET /health HTTP/1.1\" 200 353 \"-\" \"GoogleHC/1.0\""
"2023-09-13 16:43:17,946 [INFO] 172.16.0.82 - - [13/Sep/2023:16:43:17 +0000] \"GET /health HTTP/1.1\" 200 353 \"-\" \"GoogleHC/1.0\""
"2023-09-13 16:43:17,946 [INFO] 172.16.0.82 - - [13/Sep/2023:16:43:17 +0000] \"GET /health HTTP/1.1\" 200 353 \"-\" \"GoogleHC/1.0\""
"2023-09-13 16:43:18,555 [INFO] 172.16.0.65 - - [13/Sep/2023:16:43:18 +0000] \"GET /health HTTP/1.1\" 200 353 \"-\" \"GoogleHC/1.0\""
"2023-09-13 16:43:18,555 [INFO] 172.16.0.65 - - [13/Sep/2023:16:43:18 +0000] \"GET /health HTTP/1.1\" 200 353 \"-\" \"GoogleHC/1.0\""
"2023-09-13 16:43:18,739 [INFO] 172.16.0.28 - - [13/Sep/2023:16:43:18 +0000] \"GET /health HTTP/1.1\" 200 353 \"-\" \"GoogleHC/1.0\""
"2023-09-13 16:43:18,739 [INFO] 172.16.0.28 - - [13/Sep/2023:16:43:18 +0000] \"GET /health HTTP/1.1\" 200 353 \"-\" \"GoogleHC/1.0\""
"2023-09-13 16:43:18,757 [INFO] 172.16.0.65 - - [13/Sep/2023:16:43:18 +0000] \"GET /health HTTP/1.1\" 200 353 \"-\" \"GoogleHC/1.0\""
"2023-09-13 16:43:18,757 [INFO] 172.16.0.65 - - [13/Sep/2023:16:43:18 +0000] \"GET /health HTTP/1.1\" 200 353 \"-\" \"GoogleHC/1.0\""
"2023-09-13 16:43:19,631 [INFO] 172.16.0.16 - - [13/Sep/2023:16:43:19 +0000] \"GET /health HTTP/1.1\" 200 353 \"-\" \"GoogleHC/1.0\""
"2023-09-13 16:43:19,631 [INFO] 172.16.0.16 - - [13/Sep/2023:16:43:19 +0000] \"GET /health HTTP/1.1\" 200 353 \"-\" \"GoogleHC/1.0\""
"2023-09-13 16:43:22,621 [INFO] Handling signal: ttin"
"2023-09-13 16:43:22,621 [INFO] Handling signal: ttin"
"2023-09-13 16:43:22,658 [INFO] Booting worker with pid: 62"
"2023-09-13 16:43:22,658 [INFO] Booting worker with pid: 62"
"2023-09-13 16:43:23,247 [INFO] 172.16.0.13 - - [13/Sep/2023:16:43:23 +0000] \"GET /health HTTP/1.1\" 200 353 \"-\" \"GoogleHC/1.0\""
"2023-09-13 16:43:23,247 [INFO] 172.16.0.13 - - [13/Sep/2023:16:43:23 +0000] \"GET /health HTTP/1.1\" 200 353 \"-\" \"GoogleHC/1.0\""
"2023-09-13 16:43:23,253 [INFO] 172.16.0.99 - - [13/Sep/2023:16:43:23 +0000] \"GET /health HTTP/1.1\" 200 353 \"-\" \"GoogleHC/1.0\""
"2023-09-13 16:43:23,253 [INFO] 172.16.0.99 - - [13/Sep/2023:16:43:23 +0000] \"GET /health HTTP/1.1\" 200 353 \"-\" \"GoogleHC/1.0\""
"2023-09-13 16:43:23,489 [INFO] 172.16.0.99 - - [13/Sep/2023:16:43:23 +0000] \"GET /health HTTP/1.1\" 200 353 \"-\" \"GoogleHC/1.0\""
"2023-09-13 16:43:23,489 [INFO] 172.16.0.99 - - [13/Sep/2023:16:43:23 +0000] \"GET /health HTTP/1.1\" 200 353 \"-\" \"GoogleHC/1.0\""
"2023-09-13 16:43:23,724 [INFO] Handling signal: ttou"
"2023-09-13 16:43:23,724 [INFO] Handling signal: ttou"
"2023-09-13 16:43:23,725 [INFO] Worker exiting (pid: 44)"
"2023-09-13 16:43:23,725 [INFO] Worker exiting (pid: 44)"
"2023-09-13 16:43:23,738 [INFO] 172.16.0.99 - - [13/Sep/2023:16:43:23 +0000] \"GET /health HTTP/1.1\" 200 353 \"-\" \"GoogleHC/1.0\""
"2023-09-13 16:43:23,738 [INFO] 172.16.0.99 - - [13/Sep/2023:16:43:23 +0000] \"GET /health HTTP/1.1\" 200 353 \"-\" \"GoogleHC/1.0\""
"2023-09-13 16:43:24,789 [ERROR] Worker (pid:44) was sent SIGTERM!"
"2023-09-13 16:43:24,789 [ERROR] Worker (pid:44) was sent SIGTERM!"
```
I have discovered that `"[INFO] Worker exiting (pid: 44)"` also existed in previous versions of Airflow, indicating it has probably always been present in our Kubernetes deployment. Given this context, my primary inquiry is: Can we confidently disregard this error in our logs?

**Reply by potiuk** [reply]
I think you need to see at your health check logs - do they send the SIGTERM because they think that your gunicorns are not healthy ? can you find that out by checking your kubernetes probes that are running the healthchecks? There are ways you can check it in your kubernetes deployment - depending on your k8s installation (I guess you can google it in case you've never done it - but I think it's good to learn how to do it for any other probes). What do the probes report?

**Reply by VladimirYushkevich** [reply]
Sorry, I didn't find any health check probes reports. But we figured out that those signal correlates with [worker_refresh_interval](https://airflow.apache.org/docs/apache-airflow/2.7.1/configurations-ref.html#worker-refresh-interval)(default is 100min). My next questions are why it goes to ERROR now and can I disable those check at all by setting `worker_refresh_batch_size=0`?

**Reply by potiuk** [reply]
> My next questions are why it goes to ERROR now 

I have absolutely no idea. But might be worth now re-creating an issue at gunicorn possibly ? . This error seems to be printed by gunicorn: https://github.com/benoitc/gunicorn/blob/ab9c8301cb9ae573ba597154ddeea16f0326fc15/gunicorn/arbiter.py#L548  - maybe there is a way to configure gunicorn to disable it ?

Seems that it is like that for at least 3 years in gunicorn https://github.com/benoitc/gunicorn/commit/28df9926d7c650248d267eeafaa8b2c2cdd9af38

> and can I disable those check at all by setting worker_refresh_batch_size=0?

```
    :param worker_refresh_batch_size: Number of workers to refresh at a time. When set to 0, worker
        refresh is disabled. When nonzero, airflow periodically refreshes webserver workers by
        bringing up new ones and killing old ones.
```

So if this only about this refreshing, I woudl recommend to keep it and ignore the error. Periodic refreshing of the worker is good as it allows you to avoid side-effects and growing memory.

**Reply by VladimirYushkevich** [reply, marked as answer]
We are hitting exactly the same [issue](https://github.com/benoitc/gunicorn/issues/3050). I revisited our requirements.txt after upgrade and indeed for `gunicorn` we have a following change from `20.1.0` to `21.2.0`. Seems like there is a fix already from `gunicorn` to suppress above error message in log.
